subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DefaultCopySpec.java

```diff
@@ -55,30 +59,28 @@
 public class DefaultCopySpec implements CopySpecInternal {
     private static final NotationParser<Object, String> PATH_NOTATION_PARSER = PathNotationConverter.parser();
     protected final FileResolver fileResolver;
-    private final Set<Object> sourcePaths;
+    private final Set<Object> sourcePaths = new LinkedHashSet<Object>();
     private Object destDir;
     private final PatternSet patternSet;
-    private final List<CopySpecInternal> childSpecs;
-    private final List<CopySpecInternal> childSpecsInAdditionOrder;
+    private final List<CopySpecInternal> childSpecs = new LinkedList<CopySpecInternal>();
```

Comment:
Why did you change `childSpecs`, `childSpecsInAdditionOrder` and `copyActions` from `ArrayList` to `LinkedList`?

Reply:
Because there are typically no child specs or a single or a couple child specs. `ArrayList` optimizes for the no child spec case, but goes on to allocate 10 entries as soon as you add something. I expect a `LinkedList` is more efficient for these cases, e.g. the single child most root specs have.